### PR TITLE
refactor: splitting off  `AttestationSigners`

### LINKED ISSUE #41: Consider splitting `AllocationMonitor` from `AttestationSigners`, etc
Would it make sense?
I'm considering this because TAP will also need to watch for new allocations, and having lots of TAP stuff inside of `AllocationMonitor` may be a little bit of an anti-pattern.
For such a change, I was looking into leveraging `tokio::sync::broadcast` or `tokio::sync::watch`, and the other components would just have to subscribe to know when the current set of eligible allocations has changed.

That may be overkill though :shrug:

### PR DESCRIPTION
**Pull request by aasseman**
Now can subscribe to `AllocationMonitor` updates, so `AttestationSigners` knows when to update its signers list. Makes way for TAP to also watch the `AllocationMonitor`.

Note: Moved some test vectors out of allocation_monitor.rs to share them amongst test files.

Fixes #41

### CODE REVIEW

File: service/src/allocation_monitor.rs [hunk 1]
```diff
@@ -579,7 +369,8 @@ mod tests {
             std::env::var("NETWORK_SUBGRAPH_ID").expect("NETWORK_SUBGRAPH_ID not set");
         let indexer_address = std::env::var("INDEXER_ADDRESS").expect("INDEXER_ADDRESS not set");
         // If you don't specify the correct mnemonic, the attestation signers won't be found
-        let indexer_mnemonic = std::env::var("INDEXER_MNEMONIC").expect("INDEXER_MNEMONIC not set");
+        let _indexer_mnemonic =
```
Comment (hopeyen) [inline service/src/allocation_monitor.rs:372]: Can this be removed?

File: service/src/allocation_monitor.rs [hunk 2]
```diff
@@ -24,44 +21,42 @@ struct AllocationMonitorInner {
...
-    dispute_manager: Address,
+    watch_sender: Sender<()>,
+    watch_receiver: Receiver<()>,
 }
 
+#[cfg_attr(test, faux::create)]
 #[derive(Debug, Clone)]
 pub struct AllocationMonitor {
-    monitor_handle: Arc<tokio::task::JoinHandle<()>>,
+    _monitor_handle: Arc<tokio::task::JoinHandle<()>>,
```
Comment (hopeyen) [inline service/src/allocation_monitor.rs:31]: what do we expect this to do? 

Is it right to think of `AllocationMonitorInner` as something that just keeps information including watcher and receiver while `AllocationMonitor` is doing the actual monitoring?

File: service/src/attestation_signers.rs [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+    ) -> Self {
+        let inner = Arc::new(AttestationSignersInner {
+            attestation_signers: Arc::new(RwLock::new(HashMap::new())),
+            allocation_monitor,
+            indexer_mnemonic,
+            chain_id,
+            dispute_manager,
+        });
+
+        let _update_loop_handle = {
```
Comment (hopeyen) [inline service/src/attestation_signers.rs:46]: Similar question here on why do we keep _update_loop_handle for attestation signers?

File: service/src/allocation_monitor.rs (hunk 1 above)
Comment (aasseman) [reply service/src/allocation_monitor.rs:372]: Yes! Actually I don't remember adding the underscore there, weird.

Comment (aasseman) [reply service/src/allocation_monitor.rs]: Fix in a410943128cabe8166880b041bb287714e770db0

File: service/src/allocation_monitor.rs (hunk 2 above)
Comment (aasseman) [reply service/src/allocation_monitor.rs:31]: The main reason I created an "inner" is to make this possible:
https://github.com/graphops/indexer-service-rs/blob/a410943128cabe8166880b041bb287714e770db0/service/src/allocation_monitor.rs#L58-L63

Basically I needed the spawned thread not depend on `AllocationMonitor`, because it is itself belonging to `AllocationMonitor`.
I could just clone all the values that the thread needs, but the "inner" makes it visually cleaner. Also makes it cleaner to Arc clone `AllocationMonitor`.

tl;dr it's mostly easthetics, I guess

Comment (aasseman) [reply service/src/allocation_monitor.rs:31]: As for the handle, I don't like the idea of not explicitely keeping it somewhere. Is the thread even guaranteed to continue existing if the handle is lost? (I actually don't know yet what's going on at a lower level here)

Comment (hopeyen) [reply service/src/allocation_monitor.rs:31]: And you are keeping track of the handle so it won't "roam free"? Would we need to specifically shutdown the spawned threads when we get termination or exit signal?

Comment (aasseman) [reply service/src/allocation_monitor.rs:31]: > Would we need to specifically shutdown the spawned threads when we get termination or exit signal?

Well, that's a good question. I guess we'll see when we tie everything together :grin:

Comment (aasseman) [reply service/src/allocation_monitor.rs:31]: Calling the Rust masters! Would @axiomatic-aardvark know?

Comment (axiomatic-aardvark) [reply service/src/allocation_monitor.rs:31]: I think it makes sense either way, the thread will continue even without the handle, but it does seem more idiomatic to keep track of it. And as @hopeyen suggested that allows for a shutdown signal (maybe a `Arc<AtomicBool>`) to be used to break out of the loop

**Review by hopeyen — APPROVED** [review]
coolio! we can find out what happens when the components get all tied together
